File: src/main/java/org/taktik/icure/logic/impl/filter/service/ServiceByHcPartyTagCodeDateFilter.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic.impl.filter.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.taktik.icure.entities.User;
import org.taktik.icure.entities.embed.Service;
import org.taktik.icure.logic.ContactLogic;
import org.taktik.icure.logic.ICureSessionLogic;
import org.taktik.icure.logic.SessionLogic;
import org.taktik.icure.logic.impl.filter.Filter;
import org.taktik.icure.logic.impl.filter.Filters;

import javax.security.auth.login.LoginException;
import java.util.HashSet;
import java.util.List;
import java.util.Set;


public class ServiceByHcPartyTagCodeDateFilter implements Filter<String, Service, org.taktik.icure.dto.filter.service.ServiceByHcPartyTagCodeDateFilter> {
	ContactLogic contactLogic;
	ICureSessionLogic sessionLogic;

	@Autowired
	public void setContactLogic(ContactLogic contactLogic) {
		this.contactLogic = contactLogic;
	}
	@Autowired
	public void setSessionLogic(ICureSessionLogic sessionLogic) {
		this.sessionLogic = sessionLogic;
	}

	@Override
	public Set<String> resolve(org.taktik.icure.dto.filter.service.ServiceByHcPartyTagCodeDateFilter filter, Filters context) {
		try {
            String hcPartyId = filter.getHealthcarePartyId() != null ? filter.getHealthcarePartyId() : getLoggedHealthCarePartyId();
            HashSet<String> ids = null;
            if (filter.getTagType() != null && filter.getTagCode() != null) {
                ids = new HashSet<>(contactLogic.findServicesByTag(
                        hcPartyId,
                        filter.getPatientSecretForeignKey(), filter.getTagType(),
                        filter.getTagCode(), filter.getStartValueDate(), filter.getEndValueDate()
                ));
            }

            if (filter.getCodeType() != null && filter.getCodeCode() != null) {
                List<String> byCode = contactLogic.findServicesByCode(
                        hcPartyId,
                        filter.getPatientSecretForeignKey(), filter.getCodeType(),
                        filter.getCodeCode(), filter.getStartValueDate(), filter.getEndValueDate()
                );
                if (ids==null) { ids = new HashSet<>(byCode); } else { ids.retainAll(byCode); }
            }

            return ids != null ? ids : new HashSet<>();
		} catch (LoginException e) {
			throw new IllegalArgumentException(e);
		}
	}

	private String getLoggedHealthCarePartyId() throws LoginException {
		User user = sessionLogic.getCurrentSessionContext().getUser();
		if (user == null || user.getHealthcarePartyId() == null) {
			throw new LoginException("You must be logged to perform this action. ");
		}
		return user.getHealthcarePartyId();
	}
}

File: src/main/kotlin/org/taktik/icure/asynclogic/impl/filter/service/ServiceByHcPartyTagCodeDateFilter.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.asynclogic.impl.filter.service

import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.toSet
import org.taktik.icure.asynclogic.AsyncSessionLogic
import org.taktik.icure.asynclogic.ContactLogic
import org.taktik.icure.asynclogic.impl.filter.Filter
import org.taktik.icure.asynclogic.impl.filter.Filters
import org.taktik.icure.domain.filter.service.ServiceByHcPartyTagCodeDateFilter
import org.taktik.icure.entities.embed.Service
import org.taktik.icure.utils.getLoggedHealthCarePartyId
import java.util.*
import javax.security.auth.login.LoginException

class ServiceByHcPartyTagCodeDateFilter(private val contactLogic: ContactLogic,
                                        private val sessionLogic: AsyncSessionLogic) : Filter<String, Service, ServiceByHcPartyTagCodeDateFilter> {
    override fun resolve(filter: ServiceByHcPartyTagCodeDateFilter, context: Filters) = flow {
        try {
            val hcPartyId = filter.healthcarePartyId ?: getLoggedHealthCarePartyId(sessionLogic)
            var ids: HashSet<String>? = null
            val patientSFK = filter.patientSecretForeignKey
            val patientSFKList = if (patientSFK != null) listOf(patientSFK) else null
            if (filter.tagType != null && filter.tagCode != null) {
                ids = HashSet(contactLogic.listServiceIdsByTag(
                        hcPartyId,
                        patientSFKList, filter.tagType!!,
                        filter.tagCode!!, filter.startValueDate, filter.endValueDate
                ).toSet())
            }
            if (filter.codeType != null && filter.codeCode != null) {
                val byCode = contactLogic.listServiceIdsByCode(
                        hcPartyId,
                        patientSFKList, filter.codeType!!,
                        filter.codeCode!!, filter.startValueDate, filter.endValueDate
                ).toSet()
                if (ids == null) {
                    ids = HashSet(byCode)
                } else {
                    ids.retainAll(byCode)
                }
            }
            emitAll((ids ?: HashSet()).asFlow())
        } catch (e: LoginException) {
            throw IllegalArgumentException(e)
        }
    }
}
